### PR TITLE
Merge dev branch: fix noise generation bugs, adds fast perfect CRHWP demodulation. Fixes map2tod to combine multiple systematics consistently. 

### PR DESCRIPTION
**Pull request by gfabbian**
Fixes missing sampling frequency factor to convert uk*sqrt(s) noise in noise per sample.
Fixes flag for gdeprojection factor and API
Fixes bug in CRWHP mapmaking that lead to mapping of only a single bolometer in a pair.
Adds flag in map2tod to create perfectly demodulated timestreams (delta function line filter) directly when doing map2tod. removes the need of demodulation and computational cost.
Restructure codes to be able to combine boresight pointing systematics with differential pointing/beam systematics and pol. angle systematics.

### CODE REVIEW

File: s4cmb/tod.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+# !/usr/bin/python
```
Comment (JulienPeloton) [inline s4cmb/tod.py:1]: I do not  think there should be a space here (probably a wrong formatting from  the text editor)

File: s4cmb/tod.py [hunk 2]
```diff
@@ -1060,88 +1096,136 @@ def map2tod(self, ch):
...
+            sin2pol_ang_in = np.sin(2 * pol_ang_in)
 
             # Use pa of pair center for tod2map if differential pointing
-            # is used. Otherwise the polarization angle of a pair is the
-            # same as the top and bottom detectors.
-            # TODO: I do not understand this try/except.
-            # TODO: remove the bare exception!!!
+            # is used (and thus the pa_pair is defined). Throws an exception if
+            # this is not the case and uses the same angle for the
+            # the top and bottom detectors.
```
Comment (JulienPeloton) [inline s4cmb/tod.py:1112]: Good catch!

File: s4cmb/tod.py [hunk 3]
```diff
@@ -1363,6 +1445,7 @@ def __init__(
         array_noise_level2=None,
         array_noise_seed2=56736,
         mapping_perpair=False,
+        bandpass=1.9,
```
Comment (JulienPeloton) [inline s4cmb/tod.py:1448]: Where this 1.9 is coming from?

File: s4cmb/tod.py [hunk 4]
```diff
@@ -1408,6 +1491,10 @@ def __init__(
             If True, assume that you want to process pairs of bolometers
             one-by-one, that is pairs are uncorrelated. Default is False (and
             should be False unless you know what you are doing).
+        bandpass : float, optional
+            Band width around your frequency of interest [Hz] defined relative
+            to the spinning frequency of the HWP. If not specified it is equal
+            to the HWP spinning frequency itself. Default is 1.9.
```
Comment (JulienPeloton) [inline s4cmb/tod.py:1497]: You said `If not specified it is equal to the HWP spinning frequency itself`, but the default value (i.e. nothing specified) is 1.9. What do you mean then?

File: s4cmb/tod_f.f90 [hunk 5]
```diff
@@ -180,6 +183,19 @@ subroutine tod2map_hwp_f(d0, d4r, d4i, w0, w4, nhit, waferi1d, &
                     d0(pixel) = d0(pixel)+ waferts(if0) * weight0(j)
                     d4r(pixel) = d4r(pixel) + (c*waferts(i4r)+s*waferts(i4i)) * weight4(j)
                     d4i(pixel) = d4i(pixel) + (s*waferts(i4r)-c*waferts(i4i)) * weight4(j)
+
+                    ! second detector of a focal plane pixel (rotated of pi/2)
```
Comment (JulienPeloton) [inline s4cmb/tod_f.f90:187]: Good catch!

**Review by JulienPeloton — APPROVED** [review]
Thanks @gfabbian ! Sorry for the long delay... The new code is great, I left some minor comments for my understanding.

File: s4cmb/tod.py (hunk 3 above)
Comment (gfabbian) [reply s4cmb/tod.py:1448]: It was the previous hardcoded value at line 1486

File: s4cmb/tod.py (hunk 4 above)
Comment (gfabbian) [reply s4cmb/tod.py:1497]: that's an error. Should be : If set to None, it is equal to the HWP spinning frequency itself.

Comment (gfabbian) [reply s4cmb/tod.py:1497]: I've pushed it and fixed it in the last commit.

File: s4cmb/tod.py (hunk 3 above)
Comment (JulienPeloton) [reply s4cmb/tod.py:1448]: Hum, I do not remember at all why this value... But that's me  indeed that introduced it. Probably taken from PB pipelines at that time :D

File: s4cmb/tod.py (hunk 4 above)
Comment (JulienPeloton) [reply s4cmb/tod.py:1497]: Thanks!

**Comment by JulienPeloton** [comment]
Thanks Gigi, I am merging.

**Pull request merged**